README.md
java_wrapper Cookbook
=====================

### Status
[![Build Status](https://travis-ci.org/gsaslis/chef-java-wrapper.svg?branch=master)](https://travis-ci.org/gsaslis/chef-java-wrapper)

Overview
========
This cookbook provides `java_wrapper`, a resource that wrap a java application
to be easily managed using start up scripts. It uses http://wrapper.tanukisoftware.com/.

Requirements
============
This cookbook has only be tested on linux system. Tanuki java wrapper works on
Windows system, but considering this cookbook has never been tested on Windows,
it most likely not work.

Attribute Parameters
----------
Customize the attributes to suit your need, the following attributes
have to be provided:

- `app_name`: the name of the application, if not provided, it uses the name of the resource.
- `classpath`: the classpath of the application
- `app_parameters`: the array of String that will be passed to the application
- `java_parameters`: The array of java parameters
- `log_file_name`: the file name where the logs will be stored. Use something relevant to your app_name, to make it clear what log you're readind.
- `wrapper_working_dir`: configuring the Working Directory.

This attributes can be used to further customize the application:
- `app_long_name`: the long name used for the daemon
- `app_dir`: the directory in which the application is deployed, it defaults to `/opt/#{app_name}`
- `bin_dir`: the bin directory for the wrapper, it defaults to `#{app_dir}/bin`
- `lib_idr`: the lib directory for the wrapper, it defaults to `#{app_dir}/lib`
- `conf_dir`: the conf directory for the wrapper, it defaults to `#{app_dir}/conf`
- `daemonize`: Spawns a fork of the Wrapper process as a detached non-session group leader. This means that the process will continue to run even after the console from which the Wrapper process was launched has been closed.
- `exit_timeout`: the timeout the wrapper allows to the app for termination, in seconds. (defaults to 15).
- `library_path`: an array of paths, where the wrapper will look for (Wrapper.DLL or libwrapper.so)
- `logs_dir`: the log directory for the wrapper, it defaults to `/var/log/#{name}`
- `init_mem_MB`: the mininum memory in MB for the JVM, defaults to 3
- `max_mem_MB`: the maximum memory in MB for the JVM, defaults to 64
- `native_library_dest_dir`: convenience parameter for copying the native library to this folder
- `console_flush`: Configures the output sent to the console to be completely flushed (TRUE / FALSE). False by default.
- `permissions_owner`: the system user who will be the owner of directories and files.
- `permissions_group`: the system group to which the directories and files will belong.
- `run_as_user`: the user running the wrapper, defaults to root
- `startup_timeout`: the timeout the wrapper allows between the time that the Wrapper launches the JVM process and the time that the JVM side of the Wrapper responds that the application has started. Setting this property value to "0" (zero) means never time out. The value is in seconds. (defaults to 30).
- `wrapper_version`: the version of the java wrapper, defaults to 3.5.21
- `wrapper_os`: the os for the java wrapper, defaults to linux
- `wrapper_cpu`: the cpu for the java wrapper, defaults to x86
- `wrapper_bit`: the architecture for the java wrapper, defaults to 64
- `wrapper_extension`: the extension for the java wrapper, defaults to tar.gz
- `wrapper_url`: the url to download the wrapper, defaults to "http://wrapper.tanukisoftware.com/download/#{wrapper_version}/wrapper-#{wrapper_os}-#{wrapper_cpu}-#{wrapper_bit}-#{wrapper_version}.#{wrapper_extension}"
- `file_logging_enabled`: cause file logging to be disabled if set to false. Boolean. True by default.

Usage
-----
You will find usage in the test recipe at https://github.com/arussel/chef-java-wrapper/blob/master/test/fixtures/cookbooks/java_wrapper_test/recipes/default.rb

java_wrapper 'jetty' do
  app_parameters ["org.mortbay.start.Main"]
  classpath ["/usr/local/jetty/start.jar"]
  java_parameters ["-Djetty.home=/usr/local/jetty"]
end

java_wrapper 'play' do
  classpath ["#{helloworld_dir}/target/staged/*"]
  app_parameters ["play.core.server.NettyServer", "#{helloworld_dir}"]
  java_parameters ["-Dhttp.port=9001"]
end

Contributing
------------
1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write you change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
Authors: Alexandre Russel, Yorgos Saslis


CHANGELOG.md
java-wrapper CHANGELOG
======================

This file is used to list changes made in each version of the java-wrapper cookbook.

1.2.1
-----
- Fixed Travis Build

1.2.0
-----
- Added cookbook attribute that disables file logging ( arussel/chef-java-wrapper#10 )
- Fixed warning about version incompatibility (arussel/chef-java-wrapper#9)

1.1.0
-----
- Added LWRP parameter for wrapper.console.flush

1.0.1
-----
- Fixed broken tests (due to changed URL for jetty download)
- Fixed native_library extraction feature (wrong predicate)
- Added chefspec unit tests

1.0.0
-----
- Build passes on Travis CI, using Test Kitchen with kitchen-ec2 driver for IT tests using serverspec.
- fixed rubocop and foodcritic warnings

0.6.0
-----
- Made test kitchen pass again, on both ubuntu 12.04 and ubuntu 14.04
- separated enable & start actions for resource
- Fixed missing run_as_user attribute

0.5.0
-----
- @gsaslis - Added configurable exit.timeout and startup.timeout. Also added .use_system_time=false and .request_thread_dump_on_failed_jvm_exit=TRUE as hard-coded values for now.

0.4.0
-----
- @gsaslis - Added support for wrapper.daemonize and parameterizable console name

0.3.0
-----
- @gsaslis - Bugfix: remove hard-coded path of /etc/init.d as location of executable wrapper file
- @gsaslis - Add new capability of being able to extract native library as well upon setting an extra option


0.2.0
-----
- @gsaslis - Separated Create & Enable Actions, so you can now create, without enabling the service.
- @gsaslis - Added a customizable log file name
- @gsaslis - The wrapper.conf file name will now match your app name, so you can have multiple instances of the java wrapper configs, even in the same directory.
- @gsaslis - The wrapper_url ignored version, so I've made it parameterizable, meaning you can now choose what wrapper version to install.
- @gsaslis - Added owner and group attributes for filesystem permissions to the files & dirs


0.0.1
-----
- arussel - Initial release of java-wrapper

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


spec/spec_helper.rb
# spec_helper.rb
require 'chefspec'
require 'chefspec/berkshelf'
ChefSpec::Coverage.start!

# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# The `.rspec` file also contains a few flags that are not defaults but that
# users commonly want.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end
end


spec/unit/recipes/default_spec.rb
require 'chefspec'
require 'chefspec/berkshelf'

describe 'java_wrapper::default' do
  let(:chef_run) { ChefSpec::SoloRunner.new.converge(described_recipe) }

  # it 'create directory structure' do
  #   expect(chef_run).to create_directory('/opt/java-wrapped-app/bin')
  # end
end


spec/unit/recipes/native_lib_spec.rb
require 'chefspec'
require 'chefspec/berkshelf'

describe 'java_wrapper_test::native_lib' do
  # context 'On ' do
  let(:chef_run) do
    ChefSpec::SoloRunner.new(step_into: ['java_wrapper']) do |node|
      node.set['a']['var']['here'] = ''
    end.converge(described_recipe)
  end

  it 'extracts native library' do
    expect(chef_run).to cherry_pick_ark('java_wrapper_native_lib').with(
      url: 'http://wrapper.tanukisoftware.com/download/3.5.21/wrapper-linux-x86-64-3.5.21.tar.gz',
      action: [:cherry_pick],
      path: '/usr/local/java_wrapper/lib',
      creates: 'wrapper-linux-x86-64-3.5.21/lib/libwrapper.so'
    )
  end
  # end
end


resources/default.rb
actions :create, :enable, :start, :remove
default_action :create

attribute :app_name, kind_of: String
attribute :app_long_name, kind_of: String
attribute :app_dir, kind_of: String
attribute :bin_dir, kind_of: String
attribute :lib_dir, kind_of: String
attribute :conf_dir, kind_of: String
attribute :logs_dir, kind_of: String
attribute :log_file_name, kind_of: String
attribute :init_mem_MB, kind_of: String, default: '3'
attribute :max_mem_MB, kind_of: String, default: '64'
attribute :java_parameters, kind_of: Array, default: []
attribute :classpath, kind_of: Array, default: []
attribute :app_parameters, kind_of: Array, default: []
attribute :run_as_user, kind_of: String, default: 'root'
attribute :permissions_owner, kind_of: String, default: 'root'
attribute :permissions_group, kind_of: String, default: 'root'
attribute :java_wrapper_dir, kind_of: String, default: 'java_wrapper'
attribute :wrapper_working_dir, kind_of: String
attribute :wrapper_version, kind_of: String, default: '3.5.21'
attribute :wrapper_os, kind_of: String, default: 'linux'
attribute :wrapper_cpu, kind_of: String, default: 'x86'
attribute :wrapper_bit, kind_of: String, default: '64'
attribute :wrapper_extension, kind_of: String, default: 'tar.gz'
attribute :wrapper_url, kind_of: String
attribute :daemonize, kind_of: String, default: 'true'
attribute :exit_timeout, kind_of: String, default: '15'
attribute :startup_timeout, kind_of: String, default: '30'
attribute :native_library_dest_dir, kind_of: String, default: ''    # convenience of copying native library to provided folder
attribute :console_flush, kind_of: String, default: 'FALSE'

def initialize(*args)
  super
  @app_name ||= @name
  @app_dir ||= "/opt/#{@name}"
  @bin_dir ||= "#{@app_dir}/bin"
  @lib_dir ||= "#{@app_dir}/lib"
  @conf_dir ||= "#{@app_dir}/conf"
  @logs_dir ||= "/var/log/#{@name}"
end
